data_model/src/events/data/filters.rs

```diff
@@ -1,192 +1,466 @@
+                (ByMetadataRemoved, MetadataRemoved(_)) => true,
+                (ByTotalQuantityChanged, TotalQuantityChanged(_)) => true,
+                _ => false,
+            }
+        } else {
+            true
         }
     }
 }
 
-impl<T: HasOrigin> OriginFilter<T>
-where
-    <T::Origin as Identifiable>::Id:
-        Debug + Clone + Eq + Ord + Decode + Encode + Serialize + IntoSchema,
-{
-    /// Get the id of the origin of the data event that this filter accepts.
-    pub fn origin_id(&self) -> &<T::Origin as Identifiable>::Id {
-        &self.0
+#[cfg(feature = "transparent_api")]
+impl super::EventFilter for TriggerEventFilter {
+    type Event = super::TriggerEvent;
+
+    fn matches(&self, event: &Self::Event) -> bool {
+        use TriggerEventMatcher::*;
+
+        use super::TriggerEvent::*;
+
+        if let Some(id_matcher) = &self.id_matcher {
+            if id_matcher != event.origin_id() {
+                return false;
+            }
+        }
+        if let Some(event_matcher) = &self.event_matcher {
+            match (event_matcher, event) {
+                (ByCreated, Created(_)) => true,
+                (ByDeleted, Deleted(_)) => true,
+                (ByExtended, Extended(_)) => true,
+                (ByShortened, Shortened(_)) => true,
+                _ => false,
+            }
+        } else {
+            true
+        }
     }
 }
 
 #[cfg(feature = "transparent_api")]
-impl<T: HasOrigin> Filter for OriginFilter<T>
-where
-    <T::Origin as Identifiable>::Id:
-        Debug + Clone + Eq + Ord + Decode + Encode + Serialize + IntoSchema,
-{
-    type Event = T;
-
-    fn matches(&self, event: &T) -> bool {
-        event.origin_id() == self.origin_id()
+impl super::EventFilter for RoleEventFilter {
+    type Event = super::RoleEvent;
+
+    fn matches(&self, event: &Self::Event) -> bool {
+        use RoleEventMatcher::*;
+
+        use super::RoleEvent::*;
+
+        if let Some(id_matcher) = &self.id_matcher {
+            if id_matcher != event.origin_id() {
+                return false;
+            }
+        }
+        if let Some(event_matcher) = &self.event_matcher {
+            match (event_matcher, event) {
+                (ByCreated, Created(_)) => true,
+                (ByDeleted, Deleted(_)) => true,
+                (ByPermissionRemoved, PermissionRemoved(_)) => true,
+                _ => false,
+            }
+        } else {
+            true
+        }
     }
 }
 
-impl<T: HasOrigin> PartialEq for OriginFilter<T>
-where
-    <T::Origin as Identifiable>::Id:
-        Debug + Clone + Eq + Ord + Decode + Encode + Serialize + IntoSchema,
-{
-    fn eq(&self, other: &Self) -> bool {
-        self.0 == other.0
+#[cfg(feature = "transparent_api")]
+impl EventFilter for DataEventFilter {
+    type Event = DataEvent;
+
+    fn matches(&self, event: &DataEvent) -> bool {
+        use DataEvent::*;
+        use DataEventFilter::*;
+
+        match (self, event) {
```

Comment:
What do you think about using full match on event here (without wildcard)? It would be more verbose, but it should prevent cases when someone has added new `DataEvent` and forgot to update filter at compile time.

Reply:
Yeah, I think it's a good idea. I already have managed to make this mistake once, so it's worth the boilerplate

Reply:
Not applicable anymore with multi-event matchers

Reply:
Whoops, I think it still is. It's the `EventMatcher`'s that are gone now, but not this match

Reply:
Pushed my attempt at this.

It will produce a compile error in case a new top-level data event is introduced, but will not catch new Domain event types (like Account, Asset, AssetDefinition). I think achieving that will require listing all (even non-hierarchical) event types, which is probably too much boilerplate